niceman/utils.py

```diff
@@ -888,17 +890,51 @@ def _path_(p):
         return p
 
 
-def unicode(s):
+def is_unicode(s):
+    """Return true if an object is unicode"""
+    return  (six.PY3 and isinstance(s, str)) or \
+            (six.PY2 and isinstance(s, unicode))
+
+
+def to_unicode(s, encoding="utf-8"):
     """Given a str type, convert to unicode"""
-    if PY3:
-        if isinstance(s, bytes):
-            return str(s, 'utf-8')
-        elif isinstance(s, str):
-            return s
+    # If unicode, just return it
+    if is_unicode(s):
+        return s
+    try:
+        if six.PY3:
+            return str(s, encoding)
         else:
-            raise TypeError("Incorrect type for unicode()")
+            return __builtin__.unicode(s, encoding)
+    except TypeError:
+        raise TypeError("Incorrect type for to_unicode()")
+
+
+def safe_decode(s, encoding="utf-8"):
+    """safe_decode attempts to call "decode" with the given encoding."""
+    try:
+        return s.decode(encoding=encoding)
+    except AttributeError:
+        return s
+
+
+def safe_encode(s, encoding="utf-8"):
+    """safe_encode attempts to call "encode" with the given encoding."""
+    try:
+        return s.encode(encoding=encoding)
+    except AttributeError:
+        return s
+
+
+def safe_write(os, s, encoding="utf-8"):
+    """safe_write safely writes different string types to an output stream"""
+    if PY3:
+        os.write(to_unicode(s))
     else:
-        return __builtin__.unicode(s)
+        try: # For PY2 try str first, then unicode
+            os.write(safe_encode(to_unicode(s, encoding), encoding))
```

Comment:
(I should note that it isn't *complete* madness.  to_unicode() is designed to leave existing unicode alone, so this allowed a single line to handle the case when the source was unicode or ascii.  The way to avoid the full dance and keep it in one line is to instead write a to_ascii() instead. The result would provide a nice parallel to the to_unicode() and make the purpose of the TypeError exception more obvious.)